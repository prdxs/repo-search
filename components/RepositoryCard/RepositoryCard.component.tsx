import React from 'react';
import clsx from 'clsx';
import styled from 'styled-components';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';
import CardContent from '@material-ui/core/CardContent';
import Grid from '@material-ui/core/Grid';
import StarIcon from '@material-ui/icons/Star';

import { IRepositoryCardProps } from './RepositoryCard.typings';
import { Badge } from '@material-ui/core';

const RepositoryCard: React.FC<IRepositoryCardProps> = ({
  className,
  style,
  name,
  link,
  stars,
}) => (
  <Card className={clsx('RepositoryCard-root', className)} style={style}>
    <CardHeader
      title={
        <Typography variant="h6">
          <Link href={link} target="_blank" color="secondary">
            {name}
          </Link>
        </Typography>
      }
    />
    <CardContent>
      <Grid container>
        <Grid item xs={6}>
          <Badge
            anchorOrigin={{
              vertical: 'bottom',
              horizontal: 'right',
            }}
            color="primary"
            badgeContent={stars}
            max={Infinity}
          >
            <StarIcon className="RepositoryCard-star" fontSize="large" />
          </Badge>
        </Grid>
      </Grid>
    </CardContent>
  </Card>
);

const StyledRepositoryCard = styled(RepositoryCard)`
  &.RepositoryCard-root {
    .RepositoryCard-star {
      color: #ffd700;
    }
  }
`;

export default StyledRepositoryCard;
