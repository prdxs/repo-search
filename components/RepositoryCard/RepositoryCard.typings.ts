import { IComponentProps } from '@/typings/common';

export interface IRepositoryCardProps extends IComponentProps {
  name: string;
  link: string;
  stars: number;
}
