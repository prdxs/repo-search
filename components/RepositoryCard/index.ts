export * from './RepositoryCard.typings';
export { default } from './RepositoryCard.component';
