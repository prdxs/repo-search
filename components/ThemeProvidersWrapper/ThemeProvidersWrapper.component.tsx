import React from 'react';
import { ThemeProvider } from 'styled-components';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core';
import CssBaseline from '@material-ui/core/CssBaseline';

import theme from '@/theme';
import { IComponentProps } from '@/typings/common';

const ThemeProvidersWrapper: React.FC<IComponentProps> = ({ children }) => (
  <ThemeProvider theme={theme}>
    <MuiThemeProvider theme={theme}>
      <CssBaseline />
      {children}
    </MuiThemeProvider>
  </ThemeProvider>
);

export default ThemeProvidersWrapper;
