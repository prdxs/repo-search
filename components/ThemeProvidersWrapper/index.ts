export { default } from './ThemeProvidersWrapper.component';
