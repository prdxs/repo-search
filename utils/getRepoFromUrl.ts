const getRepoFromUrl = (url: string): { owner: string; repo: string } => {
  const parts = url.split('/');
  const len = parts.length;
  const owner = parts[len - 2];
  const repo = parts[len - 1];

  return { owner, repo };
};

export default getRepoFromUrl;
