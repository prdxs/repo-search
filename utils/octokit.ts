import { Octokit } from '@octokit/rest';

const octokit = new Octokit({
  baseUrl: 'https://api.github.com',
  auth: process.env.NEXT_PUBLIC_GH_TOKEN,
});

export default octokit;
