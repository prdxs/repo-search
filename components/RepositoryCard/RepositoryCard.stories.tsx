import React from 'react';
import { withKnobs, text, number } from '@storybook/addon-knobs';

import StorybookWrapper from '@/components/StorybookWrapper';
import RepositoryCard from './RepositoryCard.component';

export default {
  title: 'RepositoryCard',
  component: RepositoryCard,
  decorators: [withKnobs],
};

export const Default: React.FC = () => {
  return (
    <StorybookWrapper>
      <RepositoryCard
        name={text('name', 'facebook/react')}
        link={text('link', 'https://github.com/facebook/react')}
        stars={number('stars', 10)}
      />
    </StorybookWrapper>
  );
};
