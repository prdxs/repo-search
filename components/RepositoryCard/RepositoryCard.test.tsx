import React from 'react';
import { mount } from 'enzyme';

import { render } from '@/test/utils';
import ThemeProvidersWrapper from '@/components/ThemeProvidersWrapper';
import RepositoryCard from './RepositoryCard.component';

describe('RepositoryCard', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <RepositoryCard
        name="prdxs/repo-search"
        link="https://github.com/prdxs/repo-search"
        stars={12}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the provided name', () => {
    const name = 'prdxs/repo-search';
    const { queryByText } = render(
      <RepositoryCard
        name={name}
        link="https://github.com/prdxs/repo-search"
        stars={12}
      />
    );

    const nameEl = queryByText(name);

    expect(nameEl).toBeInTheDocument();
  });

  it('should render the provided link', () => {
    const link = 'https://github.com/prdxs/repo-search';
    const wrapper = mount(
      <ThemeProvidersWrapper>
        <RepositoryCard name="prdxs/repo-search" link={link} stars={12} />
      </ThemeProvidersWrapper>
    );

    expect(wrapper.find('a').prop('href')).toEqual(link);
  });

  it('should render the provided number of stars', () => {
    const stars = 10;
    const { getByText } = render(
      <RepositoryCard
        name="prdxs/repo-search"
        link="https://github.com/prdxs/repo-search"
        stars={stars}
      />
    );

    const nameEl = getByText(`${stars}`);

    expect(nameEl).toHaveTextContent(`${stars}`);
  });
});
