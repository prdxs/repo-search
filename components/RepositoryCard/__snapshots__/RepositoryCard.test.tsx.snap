// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryCard should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    style="padding: 2rem;"
  >
    <div
      class="MuiPaper-root MuiCard-root RepositoryCard-root RepositoryCardcomponent__StyledRepositoryCard-sc-1c2lx1b-0 gnTwgd MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorSecondary"
              href="https://github.com/prdxs/repo-search"
              target="_blank"
            >
              prdxs/repo-search
            </a>
          </h6>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <span
              class="MuiBadge-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root RepositoryCard-star MuiSvgIcon-fontSizeLarge"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <span
                class="MuiBadge-badge MuiBadge-anchorOriginBottomRightRectangle MuiBadge-colorPrimary"
              >
                12
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
